File: test/features/knowledge/domain/knowledge_dedup_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:lexi_core/features/dictionary/domain/entities/language.dart';
import 'package:lexi_core/features/knowledge/domain/entities/knowledge_note.dart';
import 'package:lexi_core/features/knowledge/domain/knowledge_dedup.dart';

KnowledgeNote _n({
  required String id,
  required String title,
  String summary = '',
  List<String> tags = const [],
  String groupId = 'en_other',
}) {
  final t = DateTime.utc(2026, 1, 1);
  return KnowledgeNote(
    id: id, title: title, summary: summary, explanation: '',
    groupId: groupId, tags: tags, targetLanguage: Language.english,
    origin: KnowledgeNoteOrigin.manual, createdAt: t, updatedAt: t,
  );
}

void main() {
  test('matches on a repeated phrase in the title', () {
    final notes = [
      _n(id: 'a', title: 'Câu điều kiện loại 2 và 3'),
      _n(id: 'b', title: 'Mệnh đề quan hệ'),
    ];
    final hits = findRelatedNotes(
      prompt: 'giải thích câu điều kiện loại 2 khi nào dùng', notes: notes);
    expect(hits.map((n) => n.id), ['a']);
  });

  test('matches cross-diacritic (folded)', () {
    final notes = [_n(id: 'a', title: 'Câu điều kiện loại 2')];
    final hits = findRelatedNotes(prompt: 'cau dieu kien loai 2', notes: notes);
    expect(hits.single.id, 'a');
  });

  test('tag exact match scores', () {
    final notes = [_n(id: 'a', title: 'Ghi chú', tags: ['toeic'])];
    final hits = findRelatedNotes(prompt: 'ôn toeic phần này', notes: notes);
    expect(hits.single.id, 'a');
  });

  test('below threshold → no match', () {
    final notes = [_n(id: 'a', title: 'Mệnh đề quan hệ')];
    final hits = findRelatedNotes(prompt: 'thì hiện tại đơn', notes: notes);
    expect(hits, isEmpty);
  });

  test('"thì" / "thể" are content words, not stopwords', () {
    final notes = [_n(id: 'a', title: 'Thì thể')];
    // The only overlap is "thi"/"the" — if they were stopwords this scores 0.
    final hits = findRelatedNotes(prompt: 'thì thể', notes: notes);
    expect(hits.single.id, 'a');
  });

  test('caps at 3, sorted by score desc', () {
    final notes = [
      _n(id: 'a', title: 'thì hiện tại đơn'),
      _n(id: 'b', title: 'thì hiện tại tiếp diễn'),
      _n(id: 'c', title: 'thì hiện tại hoàn thành'),
      _n(id: 'd', title: 'thì quá khứ đơn'),
    ];
    final hits = findRelatedNotes(
        prompt: 'thì hiện tại đơn giải thích', notes: notes);
    expect(hits.length, 3);
    // "thi" (thì) is no longer a stopword, so note a matches the whole
    // "thi hien tai don" phrase and outscores the partial-phrase matches.
    expect(hits.first.id, 'a');
  });
}

File: apps/web/src/lib/knowledgeDedup.test.ts
import { describe, expect, it } from "vitest";
import { findRelatedNotes } from "./knowledgeDedup";
import type { KnowledgeNote } from "./knowledgeNotes";

const n = (over: Partial<KnowledgeNote>): KnowledgeNote => ({
  id: "x", title: "", summary: "", explanation: "", patterns: [], examples: [], pitfalls: [],
  groupId: "en_other", tags: [], cefrLevel: null, targetLanguage: "english",
  source: "manual", sourcePrompt: null, createdAt: "2026-01-01T00:00:00.000Z",
  updatedAt: "2026-01-01T00:00:00.000Z", ...over,
});

describe("findRelatedNotes", () => {
  it("matches on a repeated phrase in the title", () => {
    const hits = findRelatedNotes({
      prompt: "giải thích câu điều kiện loại 2 khi nào dùng",
      notes: [n({ id: "a", title: "Câu điều kiện loại 2 và 3" }), n({ id: "b", title: "Mệnh đề quan hệ" })],
    });
    expect(hits.map((h) => h.id)).toEqual(["a"]);
  });
  it("matches cross-diacritic", () => {
    expect(findRelatedNotes({ prompt: "cau dieu kien loai 2", notes: [n({ id: "a", title: "Câu điều kiện loại 2" })] })
      .map((h) => h.id)).toEqual(["a"]);
  });
  it("tag exact match scores", () => {
    expect(findRelatedNotes({ prompt: "ôn toeic phần này", notes: [n({ id: "a", title: "Ghi chú", tags: ["toeic"] })] })
      .map((h) => h.id)).toEqual(["a"]);
  });
  it("thì / thể are content words (whole-branch fix)", () => {
    expect(findRelatedNotes({ prompt: "thì thể", notes: [n({ id: "a", title: "Thì thể" })] })
      .map((h) => h.id)).toEqual(["a"]);
  });
  it("below threshold → []", () => {
    expect(findRelatedNotes({ prompt: "hiện tại đơn", notes: [n({ id: "a", title: "Mệnh đề quan hệ" })] })).toEqual([]);
  });
  it("caps at 3, sorted by score desc", () => {
    const notes = ["hiện tại đơn", "hiện tại tiếp diễn", "hiện tại hoàn thành", "quá khứ đơn"]
      .map((t, i) => n({ id: String(i), title: t }));
    const hits = findRelatedNotes({ prompt: "hiện tại đơn giải thích", notes });
    expect(hits).toHaveLength(3);
    expect(hits[0].id).toBe("0");
  });
});
